### PR TITLE
Environment events & runtime events

### PR DESCRIPTION
**Pull request by Wiezzel**
Environment events:
  * ENV_ENABLED
  * ENV_DISABLED
  * PREREQUISITES_INSTALLED
  * CONFIG_UPDATED
  * ERROR_OCCURRED

Runtime events:
  * PREPARED
  * STARTED
  * STOPPED
  * TORN_DOWN
  * ERROR_OCCURRED

### CODE REVIEW

**Comment by Wiezzel** [comment]
@Krigpl @mfranciszkiewicz Plx, review

**Comment by Krigpl** [comment]
Is it somewhere described what will be listening on these events and how it will react?

**Comment by Wiezzel** [comment]
@Krigpl I believe no. But I can image some scenarios, e.g. forwarding these events to UI to provide feedback for the user.

**Comment by Krigpl** [comment]
I have mixed feelings about adding ~700 lines of code for a feature we're not sure how it'll be used at the moment.

File: golem/envs/__init__.py [hunk 1]
```diff
@@ -16,11 +18,41 @@
...
+class RuntimeEvent(NamedTuple):
+    type: RuntimeEventType
+    details: Optional[Dict[str, Any]] = None
+
+
+RuntimeEventListener = Callable[[RuntimeEvent], Any]
+
+
+class EnvEventType(Enum):
+    ENV_ENABLED = 1
```
Comment (mfranciszkiewicz) [inline golem/envs/__init__.py]:
```suggestion
    ENABLED = 1
```

File: golem/envs/__init__.py [hunk 2]
```diff
@@ -16,11 +18,41 @@
...
+    ENV_DISABLED = 2
```
Comment (mfranciszkiewicz) [inline golem/envs/__init__.py]:
```suggestion
    DISABLED = 2
```

File: golem/envs/__init__.py [hunk 3]
```diff
@@ -149,38 +183,78 @@ def _change_status(
...
+            details=details
+        )
+        self._logger.debug("Emit event: %r", event)
+        for listener in self._event_listeners.get(event_type, ()):
+            listener(deepcopy(event))
+
+    def _prepared(self, *_) -> None:
+        """ Acknowledge that Runtime has been prepared. Log message, set status
+            and emit event. Arguments are ignored (for callback use). """
+        self._logger.info("Runtime prepared.")
```
Comment (mfranciszkiewicz) [inline golem/envs/__init__.py:216]: Maybe include class name for more information?

**Comment by mfranciszkiewicz** [comment]
We **do** need these events to give feedback to the user.

File: golem/envs/__init__.py [hunk 4]
```diff
@@ -275,16 +348,83 @@ class Environment(ABC):
...
+            every listener registered for this type of events. """
+
+        event = EnvEvent(
+            env_id=self.metadata().id,
+            type=event_type,
+            details=details
+        )
+        self._logger.debug("Emit event: %r", event)
+        for listener in self._event_listeners.get(event_type, ()):
+            listener(deepcopy(event))
```
Comment (mfranciszkiewicz) [inline golem/envs/__init__.py]: Sending events should be synchronous but I'm not entirely sure that we need to wait for the listener to finish executing.

Comment (Wiezzel) [reply golem/envs/__init__.py]: What would you suggest then? Pushing events to a queue and having a delegated event-handling thread? Or maybe spawning a new thread for each event/handler?

File: golem/envs/__init__.py (hunk 3 above)
Comment (Wiezzel) [reply golem/envs/__init__.py:216]: I think logger name is enough (e.g. Docker CPU will use `golem.envs.docker.cpu`).

File: golem/envs/__init__.py (hunk 1 above)
Comment (Wiezzel) [reply golem/envs/__init__.py]: Done.

File: golem/envs/__init__.py (hunk 2 above)
Comment (Wiezzel) [reply golem/envs/__init__.py]: Done.

File: golem/envs/__init__.py (hunk 4 above)
Comment (mfranciszkiewicz) [reply golem/envs/__init__.py]: We can either leave it for the developer to decide how to implement the listener or delegate the call to a thread pool (e.g. via `golem.core.golem_async.async_run`)

Comment (Wiezzel) [reply golem/envs/__init__.py]: OK, resolved by calling `deferToThread`.

**Review by mfranciszkiewicz — APPROVED** [review]

**Pull request merged**